include/seqan3/contrib/parallel/reader_writer_latch.hpp

```diff
@@ -0,0 +1,323 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \brief Provides seqan3::contrib::reader_writer_latch and utility objects.
+ * \author Rene Rahn <rene.rahn AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <cassert>
+#include <mutex>
+
+#include <seqan3/contrib/parallel/latch.hpp>
+#include <seqan3/contrib/parallel/spin_delay.hpp>
+#include <seqan3/core/detail/strong_type.hpp>
+#include <seqan3/std/new>
+
+namespace seqan3::contrib
+{
+//!\cond
+//!\brief A strong type to set the writer count of a seqan3::contrib::reader_writer_latch.
+struct writer_count : public detail::strong_type<size_t, writer_count>
+{
+    //!\brief An alias for the base type.
+    using base_t = detail::strong_type<size_t, writer_count>;
+
+    //!\brief Import the base constructor.
+    using base_t::base_t;
+};
+
+//!\brief A strong type to set the reader count of a seqan3::contrib::reader_writer_latch.
+struct reader_count : public detail::strong_type<size_t, reader_count>
+{
+    //!\brief An alias for the base type.
+    using base_t = detail::strong_type<size_t, reader_count>;
+
+    //!\brief Import the base constructor.
+    using base_t::base_t;
+};
+
+/*!\brief A single-use synchronisation point for closable concurrent data structures.
+ *
+ * \details
+ *
+ * A reader writer latch is a thread coordination mechanism specifically designed to synchronise with a concurrent
+ * data structure such as the seqan3::contrib::buffer_queue. In particular, the latch is constructed with a
+ * seqan3::contrib::reader_count and a seqan3::contrib::writer_count and offers special functions for readers that
+ * arrive and writers that arrive at the synchronisation point. If all writers arrived at the synchronisation point
+ * the associated concurrent data structure is closed in a completion phase by one of the participating producer
+ * threads. Thus, the concurrent data structure will be automatically signalled that there is no more data coming.
+ * If all participating readers arrived at the synchronisation point no completion phase is invoked, but the
```

Comment:
`no completion phase is invoked`? Is that right?

Reply:
Yes, for readers we don't have a completion phase.